### PR TITLE
[-] Command generate - Add the `--schema` option and do not ask for schema by default

### PR DESCRIPTION
**Pull request by VincentMolinie**

### CODE REVIEW

File: lumber-generate.js [hunk 1]
```diff
@@ -9,7 +9,8 @@ const logger = require('./services/logger');
 
 program
   .description('Generate a backend application with an ORM/ODM configured.')
-  .option('-c, --connection-url <connectionUrl>', 'Enter the database credentials with a connection URL')
+  .option('-c, --connection-url <connectionUrl>', 'Enter the database credentials with a connection URL.')
+  .option('--schema <schema>', 'Enter your database schema.')
```
Comment (rap2hpoutre) [inline lumber-generate.js]: It seems the version without a final dot is preferred by command help. I tried `docker --help`, `npm --help`, `node --help` and `yarn --help`

File: services/prompter.js [hunk 2]
```diff
@@ -71,23 +71,7 @@ async function Prompter(program, requests) {
...
-        description: 'Leave blank by default',
-        when: answers => answers.dbDialect !== 'sqlite' && answers.dbDialect !== 'mongodb',
-        default: (args) => {
-          if (args.dbDialect === 'postgres') { return 'public'; }
-          return '';
-        },
-      });
-    }
-  }
+  envConfig.dbSchema = program.schema || process.env.DATABASE_SCHEMA || '';
```
Comment (rap2hpoutre) [inline services/prompter.js]: In [this epic](https://forest.slite.com/app/channels/-KdBU14ToWBAerCk_HGS/notes/MC6abpg~l5#a83c8ed3), we do not remove prompts. You could do something like this:

```javascript
if (isRequested('dbSchema')) {
    // TODO: Remove DATABASE_SCHEMA environment variable usage in the future major Lumber version.
    envConfig.dbSchema = program.schema || process.env.DATABASE_SCHEMA;
    if (!envConfig.dbSchema) {
      // ... prompt code here.
    }
  }
```

(the TODO part was suggested by @arnaudbesnier in one other PR)

File: services/command-generate-config-getter.js [hunk 3]
```diff
@@ -9,7 +9,6 @@ const OPTIONS_DATABASE_MANDATORY = [
   'dbPassword',
 ];
 const OPTIONS_DATABASE_OPTIONAL = [
-  'dbSchema',
```
Comment (rap2hpoutre) [inline services/command-generate-config-getter.js]: This should not be removed now (see comment below). Prompt are preserved for _now_ (no breaking change in first iteration).

File: services/database-analyzer.js [hunk 4]
```diff
@@ -127,6 +127,20 @@ function DatabaseAnalyzer(databaseConnection, config) {
     tableForeignKeysAnalyzer = new TableForeignKeysAnalyzer(databaseConnection, config);
     columnTypeGetter = new ColumnTypeGetter(databaseConnection, config.dbSchema || 'public');
 
+    if (config.dbSchema) {
+      const schemaExists = await queryInterface.sequelize
+        .query(
+          `SELECT SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = '${config.dbSchema}';`,
+          { type: queryInterface.sequelize.QueryTypes.SELECT },
+        )
+        .then(result => !!result.length);
```
Comment (rap2hpoutre) [inline services/database-analyzer.js]: Maybe `Boolean(result.length)` is more expressive (`!!result.length` could be considered as a (micro) hack). Not sure about your best practices though, so as you want.

File: services/database-analyzer.js [hunk 5]
```diff
@@ -127,6 +127,20 @@ function DatabaseAnalyzer(databaseConnection, config) {
     tableForeignKeysAnalyzer = new TableForeignKeysAnalyzer(databaseConnection, config);
     columnTypeGetter = new ColumnTypeGetter(databaseConnection, config.dbSchema || 'public');
 
+    if (config.dbSchema) {
+      const schemaExists = await queryInterface.sequelize
+        .query(
+          `SELECT SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = '${config.dbSchema}';`,
```
Comment (rap2hpoutre) [inline services/database-analyzer.js]: I don't actually know sequelize but it seems this is vulnerable to SQL injection. You could use Replacements (https://sequelize.org/master/manual/raw-queries.html#replacements). Still, I know the dev who use this query should not be the one that want to run an injection (and she actually _could_ by other ways), but I think it's better to protect SQL injection everywhere.

File: services/database-analyzer.js [hunk 6]
```diff
@@ -140,8 +154,8 @@ function DatabaseAnalyzer(databaseConnection, config) {
 
     if (_.isEmpty(schema)) {
       logger.error(
-        'Your database is empty.',
-        'Please, create some tables before running generate command.',
+        'Your database looks empty! Please create some tables before running the command.',
+        'If not, check whether you\'re using a custom database schema (use in that case the --schema option or the DATABASE_SCHEMA environment variable)',
```
[the remark this answers is not in the document]
Comment (VincentMolinie) [reply services/database-analyzer.js]: The wording has been made by @marcperrinp ;)

[the remark this answers is not in the document]
Comment (VincentMolinie) [reply services/database-analyzer.js]: You are correct about the rest though

**Comment by rap2hpoutre** [comment]
@VincentMolinie Thank you for your fix! 🎉 

There are two unresolved comments. One is not important (feel free to ignore!), but the one about SQL injection seems important to answer IMO (maybe we could fix this later, as you wish). What is your opinion?

**Comment by rap2hpoutre** [comment]
:tada:

**Pull request merged**